format/partition-statistics-spec.md

```diff
@@ -0,0 +1,43 @@
+ - the License.  You may obtain a copy of the License at
+ -
+ -   http://www.apache.org/licenses/LICENSE-2.0
+ -
+ - Unless required by applicable law or agreed to in writing, software
+ - distributed under the License is distributed on an "AS IS" BASIS,
+ - WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ - See the License for the specific language governing permissions and
+ - limitations under the License.
+ -->
+
+# Partition Statistics file format
+
+This is a specification for partition statistics files. It is designed to store statistics information 
+for every partition value as a row in the **table default format** sorted based on the `partition`.
+
+The schema of the file is as follows:
+
+| Field Id | Field Name              | Field Type   | isOptional | Doc                                                                                                                         |
+|----------|-------------------------|--------------|------------|-----------------------------------------------------------------------------------------------------------------------------|
+| 1        | partition               | StructType   | false      | See [PartitionData](https://github.com/apache/iceberg/blob/master/core/src/main/java/org/apache/iceberg/PartitionData.java) |                                                                                  
+| 2        | spec_id                 | IntegerType  | false      | partition spec id                                                                                                           |
+| 3        | data_record_count       | LongType     | false      | Count of records in data files                                                                                              |                                               
+| 4        | data_file_count         | IntegerType  | false      | Count of data files                                                                                                         |
+| 5        | pos_delete_record_count | LongType     | false      | Count of records in position delete files                                                                                   |                                                          
+| 6        | pos_delete_file_count   | IntegerType  | false      | Count of position delete files                                                                                              |                                   
+| 7        | eq_delete_record_count  | LongType     | false      | Count of records in equality delete files                                                                                   |                                  
+| 8        | eq_delete_file_count    | IntegerType  | false      | Count of equality delete files                                                                                              |                                
```

Comment:
Should these be optional since v1 tables do have them?

Reply:
Ah. Good catch. I will update it.

Reply:
Thinking more about it. Can schema be still required and values can be 0 for these in v1? Because I am not sure whether we need to define schema based on the format version. 

Let us wait for some more people's opinions on this.

Reply:
I don't have strong opinion on this. It appears to be the same from user's perspective, while optional columns help reduce storage space.

Reply:
how do we draw the line between what is optional and what is not? 

I see your have the following phases in the design:

> Phase2 implementation can be about adding more fields to the partition stats like latest sequence number for the partition, file_size_in_bytes and other metrics from the data_file spec (https://iceberg.apache.org/spec/#manifests) to support special use cases for the users.
> 
> Phase3 implementation can be for writing partition level puffin files and adding the puffin file location to the schema of partition stats. 

Do we still plan to introduce puffin files to this? Based on the discussion in the doc, it seems to have scalability concerns which I agree. So if we want to add at least NDV and bloom filter, feels like those needs to be optional stats directly in the partition stats file?

Reply:
Partition stats itself is an optional requirement for the reader. 
I believe field id 1 to 4 can be required and all others can be optional.

Reply:
I have updated the optional fields now.